ws/config.py

```diff
@@ -1,180 +1,232 @@
-            top_level = cf[self.top_level_arg]
-            if not top_level:
-                raise configargparse.ConfigFileParserException("top-level parameter '{}' not found")
-        return cf(top_level, self.subname).get_options()
-
-    def serialize(self, items):
-        """
-        Does the inverse of config parsing by taking parsed values and
-        converting them back to a string representing config file contents.
+# ==============
+# Parser classes
+# ==============
 
-        :param items: an ``OrderedDict`` with items to be written to the config file
-        :returns: contents of config file as a string
-        """
-        raise NotImplementedError
-
-    def get_syntax_description(self):
-        return ""
+class ConfigParser(configparser.ConfigParser):
+    """Drop-in replacement for :py:class:`configparser.Configparser`."""
 
-class Defaults(collections.UserDict):
-    def __init__(self):
+    def __init__(self, configfile):
         super().__init__()
-        project_name = "wiki-scripts"
-
-        # global arguments
-        config_dir = os.environ.get("XDG_CONFIG_HOME", os.path.expanduser("~/.config"))
-        self.data["config"] = os.path.join(config_dir, "{0}/{0}.conf".format(project_name))
-
-        cache_dir = os.environ.get("XDG_CACHE_HOME", os.path.expanduser("~/.cache"))
-        self.data["cache_dir"] = os.path.join(cache_dir, project_name)
-
-        # ArchWiki specifics
-        self.data["site"] = "ArchWiki"
-        # NOTE: These depend on site=ArchWiki, but (config)argparse does not support conditional defaults
-        # TODO: document the behaviour properly; also emphasize that --help does not
-        #       take defaults from the config file (which is just right)
-        self.data["api_url"] = "https://wiki.archlinux.org/api.php"
-        self.data["index_url"] = "https://wiki.archlinux.org/index.php"
-
-# strict string to bool conversion
-def argtype_bool(value):
-    if isinstance(value, bool):
-        return value
-    string = str(value).lower()
-    if string in ["1", "true", "yes"]:
+        self.configfile = configfile
+
+    def fetch_section(self, section=None, to_list=True):
+        """
+        Fetches a specific section from a config file.
+
+        :param section: section name for fetching.
+        :param to_list: defines the format of the returned value (see below).
+        :returns:
+            The data from the fetched config file section. The format is
+            a list of strings if ``to_list=True`` (default value) and
+            a dictionary otherwise.
+        """
+        with open(self.configfile) as f:
+            self.read_file(f)
+
+        if section is None:
+            section, _ = os.path.splitext(os.path.basename(sys.argv[0]))
+        if not self.has_section(section):
+            section = configparser.DEFAULTSECT
+
+        option_dict = dict(self.items(section))
+        for key, value in option_dict.items():
+            if ' ' in value:
```

Comment:
This is not good - spaces do not always mean that the value should be interpreted as a list. For example, consider spaces in file paths. See https://stackoverflow.com/questions/335695/lists-in-configparser for ideas.

Reply:
I figured it is a bad practice to use spaces in filenames, but ok, I'll investigate this problem.

Reply:
```python
if "," in value:
    option_dict[key] = [item.strip() for item in value.split(',')]
```

Now `,` must be used as a separator in config file. In command line use `'...'` to wrap filenames with the whitespaces:
```
# in cli
./script.py --files readme.md 'hello world.c' config.py

# in config file
[script]
files = readme.md, hello world.c, config.py
```

Reply:
Files can contain commas too. Using plain `str.split()` is not going to work, I'd consider combining it with `json` based on [this comment](https://stackoverflow.com/a/9735884).

Reply:
04b94d3

Such multi-value syntax is now accepted:
```
[section]
numbers1 = [1, 2, 3, 4, 5]
numbers2 = 
    [6, 7, 8, 9, 10]
files1 = [
    "README.md",
    "hello_world.py",
    "path/to/strange/file/with spaces,and, commas"
    ]
```

Multi-line value ends with the last indented line. When `ConfigParser.fetch_section()` reads the config file, all in square brackets will be converted to a Python's list of strings.